### ISSUE TITLE
DISABLED test_profiler_with_async_rpc_builtin (__main__.RpcTestWithSpawn)

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#37557]
https://app.circleci.com/pipelines/github/pytorch/pytorch/162157/workflows/2caf287f-13ef-487f-a2e6-e28381d35869/jobs/5301310/steps

```
Apr 30 02:22:58   test_profiler_with_async_rpc_builtin (__main__.RpcTestWithSpawn) ... [W record_function.cpp:97] Callbacks changed while running a record function, you might be partially overlapping a record function with a profiling scope
Apr 30 02:22:58 ERROR:root:Caught exception: 
Apr 30 02:22:58 Traceback (most recent call last):
Apr 30 02:22:58   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 187, in wrapper
Apr 30 02:22:58     fn()
Apr 30 02:22:58   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/dist_utils.py", line 80, in new_test_method
Apr 30 02:22:58     return_value = old_test_method(self, *arg, **kwargs)
Apr 30 02:22:58   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/distributed/rpc/rpc_test.py", line 895, in test_profiler_with_async_rpc_builtin
Apr 30 02:22:58     RPCExecMode.ASYNC, torch.add, args=(torch.ones(1), torch.ones(1))
Apr 30 02:22:58   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/distributed/rpc/rpc_test.py", line 844, in _profiler_test_with_rpc
Apr 30 02:22:58     rpc_event = get_function_event(events, rpc_exec_mode.value)
Apr 30 02:22:58   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/dist_utils.py", line 214, in get_function_event
Apr 30 02:22:58     event = [event for event in function_events if partial_event_name in event.name][0]
Apr 30 02:22:58 IndexError: list index out of range
Apr 30 02:22:58 exiting process with exit code: 10
Apr 30 02:22:59 Some process exited badly, terminating rest.
Apr 30 02:22:59 ERROR (1.020s)
```

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528 @jjlilley @osalpekar

**Comment by rohan-varma** [comment]
Hypothesis is that the profiling callback didn't fire in time, because we don't hold the lock when we run callbacks, so fut.wait() can return without the cb's being run. I wonder if we should build a way to ensure the profiling callback, if it exists, has been completed before returning from `fut.wait()`, otherwise we will have this race condition.

**Comment by jjlilley** [comment]
Rohan, I believe the present future behavior, where it calls notify_all (and stops the wait call) prior to executing callbacks is intentional.

IIRC, There was a discussion about it in a code review a few months back - the feeling was that it might be better to unblock execution, in case callbacks implicitly relied on the notify being sent.

So ideally we don’t change future behavior here, but maybe find another way to ensure callbacks are executed.

**Comment by rohan-varma** [comment]
@jjlilley I agree that we shouldn't change the current logic of callbacks being executed. In this case, the situation is something like:

```
with profile():
    fut = rpc_async(...)
    fut.wait()
```

The challenge is that we need to run a callback that's attached onto that future, and it needs to be completed before we exit the context manager. I guess we could make the profiler aware of futures that have been profiled, and add some private functions that validate that all these futures have had their callbacks ran, and then only use this function in tests. Performance-wise, it might be better to do this in C++ though, such as adding a private method to the future API to ensure a certain callback is ran. Although, I'm not sure if we want to clutter the future API with test-only functions.

**Comment by mrshenli** [comment]
@rohan-varma 

>  I guess we could make the profiler aware of futures that have been profiled, and add some private functions that validate that all these futures have had their callbacks ran, and then only use this function in tests. 

Can this be done on the profiler side? E.g., the profiler currently accesses the future from `rpc/api.py` before the RPC API returns the `Future` to the application. Can the callback installed by the profiler has some bound state of the current profiler and then set that state before exiting? Then in the profiler's `__exit__` function, it can check and wait on that state?

**Comment by jjlilley** [comment]
btw, if we simply moved that future's condvar.notify_all() to the end, it still won't get a guarantee that all the callbacks have been executed in the fut.wait() code snippet above.

Typically in a future-style api, the main contract is simply that the future's value has been set, rather than all the subscribers being notified and having processed the event.

**Comment by rohan-varma** [comment]
@mrshenli The rpc/api.py actually has access to the `record_function` struct that we create to profile the RPC and not a handle to the actual profiler, though yes, I think this general approach would work. The only issue is that if the profiler waits in the `__exit__` function, this introduces tight coupling with RPC code which is not ideal, and we'd have to probably add a fair amount of logic to potentially deal with failure scenarios such as in the case where futures don't complete due to an error.

**Comment by mrshenli** [comment]
@rohan-varma a wild thought, which might very likely be an overkill, is to let `add_callback` return another `Future` and mark the `Future` as complete when the callback is done :)

**Comment by mrshenli** [comment]
another occurrence: https://app.circleci.com/pipelines/github/pytorch/pytorch/164198/workflows/a9e8314e-b2d6-4fbd-8c23-0ae6a384b03c/jobs/5347468/steps

**Comment by xush6528** [comment]
You probably want a "then" API on Future, in other words, continuation callbacks on wait().

You want to do this transparently for users.
new_future = future.then(lambda value: _exit_record_function(); return value)

users is actually waiting on the new future, without noticing it's a new future, because the returned value is still the same.
new_future.wait()

**Comment by rohan-varma** [comment]
@xush6528 @mrshenli's PR to expose `add_done_callback` is somewhat related to your suggestion - however with `add_done_callback` we would still need a guarantee that the callback was actually run, which is what your suggestion would provide. I'll create an issue to discuss this.

**Comment by mrshenli** [comment]
fixed by #38352
